### ISSUE TITLE
RMail and malformed UTF-8 headers

### ISSUE THREAD

**Issue by jorrizza** [sup-heliotrope/sup#205]
In broken email messages (mostly spam) the headers aren't always encoded properly (I think?), causing sup to crash. It feeds the UTF-8 string into RMail::Header::Field.parse, which in turn defines an ASCII-8bit regex and matches the string against that. For now I've fixed it like this in rmail-sup's header.rb:

``` ruby
def parse(field)
  field = field.dup.to_str
  field.force_encoding('binary')
  if field =~ EXTRACT_FIELD_NAME_RE
```

But I've got no idea if it's sup who should send it an ASCII string, or RMail who should coerce it into an ASCII string.

And here's the traceback. The line number in header.rb is off a bit, but you get the idea:

```
/var/lib/gems/1.9.1/gems/rmail-sup-1.0.1/lib/rmail/header.rb:81:in `parse'
/var/lib/gems/1.9.1/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:233:in `block in parse_header'
/var/lib/gems/1.9.1/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:229:in `each'
/var/lib/gems/1.9.1/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:229:in `parse_header'
/var/lib/gems/1.9.1/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:194:in `parse_low'
/var/lib/gems/1.9.1/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:183:in `parse'
/var/lib/gems/1.9.1/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:331:in `parse'
/var/lib/gems/1.9.1/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:345:in `read'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/mbox.rb:84:in `block in load_message'
<internal:prelude>:10:in `synchronize'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/mbox.rb:74:in `load_message'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/message.rb:772:in `parsed_message'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/message.rb:261:in `load_from_source!'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/message.rb:360:in `build_from_source'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/poll.rb:204:in `block in poll_from'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/mbox.rb:143:in `poll'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/poll.rb:201:in `poll_from'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/poll.rb:146:in `block (2 levels) in do_poll'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/poll.rb:136:in `each'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/poll.rb:136:in `block in do_poll'
<internal:prelude>:10:in `synchronize'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/poll.rb:135:in `do_poll'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/util.rb:649:in `method_missing'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/modes/poll_mode.rb:15:in `poll'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/poll.rb:58:in `poll_with_sources'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/poll.rb:94:in `poll'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup/util.rb:649:in `method_missing'
/var/lib/gems/1.9.1/gems/sup-0.15.2/bin/sup:220:in `block (2 levels) in <module:Redwood>'
/var/lib/gems/1.9.1/gems/sup-0.15.2/lib/sup.rb:87:in `block in reporting_thread'
```

**Comment by seydar** [comment]
I can't get sup to run with 2.2.1, so i'm using ruby 2.0.0-p195. i'm getting this error as well (sup v0.22.1).

Was this bug supposed to be fixed elsewhere?

**Comment by maxmeyer** [comment]
@gauteh I'm seeing this bug as well:

~~~
--- Encoding::CompatibilityError from thread: main
incompatible encoding regexp match (ASCII-8BIT regexp with UTF-8 string)
/home/user/.gem/2.2.4/gems/rmail-sup-1.0.1/lib/rmail/header.rb:82:in `=~'
/home/user/.gem/2.2.4/gems/rmail-sup-1.0.1/lib/rmail/header.rb:82:in `parse'
/home/user/.gem/2.2.4/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:233:in `block in parse_header'
/home/user/.gem/2.2.4/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:229:in `each'
/home/user/.gem/2.2.4/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:229:in `parse_header'
/home/user/.gem/2.2.4/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:194:in `parse_low'
/home/user/.gem/2.2.4/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:183:in `parse'
/home/user/.gem/2.2.4/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:331:in `parse'
/home/user/.gem/2.2.4/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:345:in `read'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/message.rb:495:in `message_to_chunks'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/message.rb:476:in `block in message_to_chunks'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/message.rb:476:in `map'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/message.rb:476:in `message_to_chunks'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/message.rb:265:in `load_from_source!'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/message.rb:376:in `build_from_source'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/poll.rb:204:in `block in poll_from'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/maildir.rb:161:in `block in poll'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/maildir.rb:160:in `each'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/maildir.rb:160:in `each_with_index'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/maildir.rb:160:in `poll'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/poll.rb:201:in `poll_from'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/util.rb:610:in `method_missing'
/home/user/.gem/2.2.4/gems/sup-0.22.1/lib/sup/sent.rb:33:in `block in write_sent_message'
~~~

This is quite annoying as all mails are sent out triggering this bug and those mails are not saved into <sent>-mailbox. Is there a fix for that bug.

**Comment by maxmeyer** [comment]
The work around does not work for me.

**Comment by phretor** [comment]
Bump. What's the status of this? I'm keep getting it even on 0.22.1 with rbenv-installed Ruby 2.1.2 on macOS Sierra.

I see that the proposed solution seemed to be #127, but the last update on that is on Aug, 2015.

Here is my traceback, for what is worth.

```
--- Encoding::CompatibilityError from thread: poll after loading inbox
incompatible encoding regexp match (ASCII-8BIT regexp with UTF-8 string)
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rmail-sup-1.0.1/lib/rmail/header.rb:81:in `=~'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rmail-sup-1.0.1/lib/rmail/header.rb:81:in `parse'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:233:in `block in parse_header'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:229:in `each'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:229:in `parse_header'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:194:in `parse_low'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:183:in `parse'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:331:in `parse'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/rmail-sup-1.0.1/lib/rmail/parser.rb:345:in `read'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/message.rb:495:in `message_to_chunks'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/message.rb:476:in `block in message_to_chunks'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/message.rb:476:in `map'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/message.rb:476:in `message_to_chunks'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/message.rb:265:in `load_from_source!'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/message.rb:376:in `build_from_source'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/poll.rb:204:in `block in poll_from'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/maildir.rb:161:in `block in poll'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/maildir.rb:160:in `each'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/maildir.rb:160:in `each_with_index'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/maildir.rb:160:in `poll'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/poll.rb:201:in `poll_from'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/poll.rb:146:in `block (2 levels) in do_poll'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/poll.rb:136:in `each'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/poll.rb:136:in `block in do_poll'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/poll.rb:135:in `synchronize'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/poll.rb:135:in `do_poll'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/util.rb:610:in `method_missing'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/modes/poll_mode.rb:15:in `poll'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/poll.rb:58:in `poll_with_sources'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/poll.rb:94:in `poll'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup/util.rb:610:in `method_missing'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/bin/sup:207:in `block (2 levels) in <module:Redwood>'
/Users/phretor/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/gems/sup-0.22.1/lib/sup.rb:87:in `block in reporting_thread'
```

**Comment by danc86** [comment]
With sup 0.23 and rmail 1.1.4 this bug still exists although the traceback is a little different:

```
ArgumentError: invalid byte sequence in UTF-8
    /usr/share/gems/gems/rmail-1.1.4/lib/rmail/parser.rb:218:in `split'
    /usr/share/gems/gems/rmail-1.1.4/lib/rmail/parser.rb:218:in `parse_header'
    /usr/share/gems/gems/rmail-1.1.4/lib/rmail/parser.rb:194:in `parse_low'
    /usr/share/gems/gems/rmail-1.1.4/lib/rmail/parser.rb:183:in `parse'
    /usr/share/gems/gems/rmail-1.1.4/lib/rmail/parser.rb:332:in `parse'
    /usr/share/gems/gems/rmail-1.1.4/lib/rmail/parser.rb:346:in `read'
    /home/dan/sup/test/dummy_source.rb:33:in `load_message'
    /home/dan/sup/lib/sup/message.rb:807:in `parsed_message'
    /home/dan/sup/lib/sup/message.rb:268:in `load_from_source!'
    /home/dan/sup/lib/sup/message.rb:381:in `build_from_source'
    /home/dan/sup/test/test_message.rb:259:in `test_nonascii_header'
```

**Comment by danc86** [comment]
Ahh nope ignore the previous comment. That's a different error which only happens in the test suite, because it uses a different code path for loading test messages.

The actual traceback with sup 0.23 and rmail 1.1.4 is this (slightly better than before but still broken):

```
--- Encoding::CompatibilityError from thread: poll after loading inbox
incompatible encoding regexp match (ASCII-8BIT regexp with UTF-8 string)
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/address.rb:706:in `==='
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/address.rb:706:in `block in get_tokenize'
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/address.rb:650:in `loop'
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/address.rb:650:in `get_tokenize'
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/address.rb:643:in `get'
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/address.rb:402:in `address_lookahead'
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/address.rb:426:in `address'
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/address.rb:366:in `address_list'
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/address.rb:302:in `parse'
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/address.rb:789:in `parse'
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/header.rb:854:in `block in address_list_fetch'
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/header.rb:850:in `each'
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/header.rb:850:in `address_list_fetch'
/usr/share/gems/gems/rmail-1.1.4/lib/rmail/header.rb:683:in `from'
/home/dan/sup/lib/sup/message.rb:499:in `message_to_chunks'
/home/dan/sup/lib/sup/message.rb:479:in `block in message_to_chunks'
...
```

**Comment by danc86** [comment]
Okay so there are two different issues going on here, both related to non-ASCII characters in headers (which are illegal, they should be encoded using RFC2047 header encoding, but spammers don't follow the rules).

Non-ASCII bytes in the mail header itself was crashing sup, but that was fixed in rmail 1.1.2:

https://github.com/terceiro/rmail/commit/0479219aec5562735526001cfe16723d3b920ed2

If there is a nested `message/rfc822` MIME part with non-ASCII characters in its headers, that will still crash because it goes through a different code path for parsing. That's the traceback in the previous comment. Specifically, the line:

```
body.normalize_whitespace
```

wrongly forces the raw message body to be marked as UTF-8 when it should still be ASCII-8BIT because it hasn't been decoded to text yet. The `normalize_whitespace` methods the very questionable `fix_encoding!` method.
